### PR TITLE
Feat: adds TF-IDF

### PR DESCRIPTION
**Pull request by micheleriva**
This PR is a rewrite of #149. More complete info to come as soon as out of draft status.

Stages:

- [x] Preload term frequencies in documents
- [x] Calculate TF-IDF scores on search
- [x] Remove frequencies on document deletion
- [x] ~Calculate cosine similarity on TF-IDF-based scoring vectors~ to be done in the future

### CODE REVIEW

File: src/prefix-tree/node.ts [hunk 1]
```diff
@@ -32,7 +31,7 @@ export function updateParent(node: Node, parent: Node): void {
 }
 
 export function removeDocument(node: Node, docID: string): boolean {
-  const index = node.docs.indexOf(docID);
+  const index = node.docs.findIndex(doc => doc === docID);
```
Comment (ShogunPanda) [inline src/prefix-tree/node.ts:34]: Why did you change this?

File: src/utils.ts [hunk 2]
```diff
@@ -91,13 +106,45 @@ export function intersectMany<T>(arrays: T[][]): T[] {
...
-    return count === arrays.length;
-  });
+  }
+
+  return result;
+}
+
+export function insertSortedValue<T = unknown>(arr: T[], el: T, compareFn?: (a: T, b: T) => number): T[] {
+  if (!compareFn) {
+    compareFn = (a, b) => {
```
Comment (ShogunPanda) [inline src/utils.ts:122]: To allow inlining, extract this function somewhere and then reference it, rather than creating it every time.

Comment (ShogunPanda) [reply src/utils.ts:122]: Moreover, given how you use compareFn, you can insanely simplify it to:

```
function compareFn(a, b) {
  return a - b;
}
```

File: src/utils.ts [hunk 3]
```diff
@@ -107,4 +154,13 @@ export function includes<T>(array: T[] | readonly T[], element: T): boolean {
     }
   }
   return false;
-}
\ No newline at end of file
+}
+
+export function sortTokenScorePredicate(a: TokenScore, b: TokenScore): number {
+  if (a[1] > b[1]) {
```
Comment (ShogunPanda) [inline src/utils.ts:160]: Same here: you can simplify to `return a[1] - b[1]`.

File: src/utils.ts [hunk 4]
```diff
@@ -71,17 +85,18 @@ export function intersectMany<T>(arrays: T[][]): T[] {
     }
   }
 
-  const set = new Map();
+  const set: Map<string, [number, number]> = new Map();
   for (const elem of arrays[0]) {
-    set.set(elem, 1);
+    set.set(elem[0] ?? "", [1, elem[1]]);
```
Comment (ShogunPanda) [inline src/utils.ts:90]: I think this is not needed. Since the elem is in the array, how can it be `undefined`?
Same applies on line 97 and 99: the `?? ""` are not needed.

**Review by ShogunPanda — COMMENTED** [review]
Minor nits, but the rest looks great!

File: src/prefix-tree/node.ts (hunk 1 above)
Comment (micheleriva) [reply src/prefix-tree/node.ts:34]: Ups, I used to use an object, then switched back to plain string and left that behind. Thanks for noticing!

File: src/utils.ts (hunk 2 above)
Comment (micheleriva) [reply src/utils.ts:122]: Awesome thanks!

File: src/utils.ts (hunk 4 above)
Comment (micheleriva) [reply src/utils.ts:90]: Let me try :)
